design/01-proc-macro.md

```diff
@@ -0,0 +1,297 @@
+
+To use a procedural macro, a Cairo programmer will have to:
+- Declare a dependency on a package, that implements the procedural macro, by adding it to the `dependencies` field in the Scarb manifest file.
+- Use the procedural macro in Cairo code, by calling it, or adding an attribute or derive macro to a Cairo item.
+
+The overall complexity of the actual procedural macro execution should be hidden from the user abstraction level.
+Apart from requiring the Rust toolchain to be installed, the user will not have to perform any additional steps to use a procedural macro.
+In particular, this two steps can be performed without any knowledge of Rust, or even the fact that the procedural macro is implemented in Rust.
+Specifically, the following requirements will be met.
+
+#### Procedural macro packages can be used as dependencies
+
+- Scarb packages can simply declare dependency relationships on other packages that implement Cairo procedural macros.
+- Because of semantics of Scarb package resolution, it will guarantee by itself, that only one instance of given procedural macro package exists in resolved package set.
+    - In other words, Scarb will out of the box verify, that there is no simultaneous dependency on `proc-macro 1.0.0` and `proc-macro 2.0.0` or `proc-macro 1.0.1`.
+- Procedural macros will end up being actual Scarb compilation unit components, though, because they will have to be treated differently from regular components, they will not be listed under `components` fields, but rather in a new one: `plugins`.
+
+#### Procedural macro will be called from Cairo code
+
+The procedural macro have to be called from Cairo code in order to be executed during the compilation.
+In contrast to current behaviour of Cairo plugins, no longer will they be executed on each node of AST.
+
+The procedural macro will be triggered by one of three Cairo expressions
+- Macro call, e.g. `xyz!`
+- Macro attribute, e.g. `#[xyz]`
+- Macro derive, e.g. `#[derive(Xyz)]`
+
+**Example:**
+
+Scarb manifest file:
+
+```toml
+[package]
+name = "hello-macros"
+version = "0.1.0"
+
+[dependencies]
+add-macro = "0.1.0"
+tracing-macro = "0.1.0"
+to-value-macro = "0.1.0"
+```
+
+Cairo source file:
+
+```cairo
+use add_macro::add;
+use tracing_macro::instrument;
+use to_value_macro::ToValue;
+
+#[derive(ToValue)]
+struct Input {
+    value: felt252, 
+}
+
+#[instrument]
+fn main() -> felt252 {
+    let a = Input { value: 1 };
+    let b = Input { value: 2 };
+    add!(a.to_value(), b.to_value());
+}
+```
+
+### Procedural macro author perspective
+
+To implement a procedural macro, a programmer will have to:
+- Create a new package, with a `Scarb.toml` manifest file and a `src/` directory besides.
+- The manifest file must define a `cairo-plugin` target type.
+- Write a Rust library, inside the `src/` directory, that implements the procedural macro API.
+- Note, that the package does not contain any `Cargo.toml` file, as it will be generated by Scarb.
```

Comment:
Will a user be able to add needed rust packages as dependencies to the macro package?

Reply:
Yes, I think you're looking for [this](https://github.com/software-mansion/scarb/pull/1109/files#diff-fa728b0211f1a1dbe27555484d8998cb7f4dd6970c00f90171419cfb5a585643R212)

Reply:
Yes, it will be possible.